Miscellaneous end session fixes

vadz opened the pull request:
This is mostly for wxMSW (see #23940, which this still doesn't fix, but which motivated these changes), but also touches wxGTK because I wanted to have a way to get the number of currently open modal dialogs in common code and this rendered wxGTK-specific code doing the same thing unnecessary.

@vslavik I thought you might be interested in having a look at the changes here, as this is also somewhat related to your changes in #24903.

@marekr I don't think this fixes KiCad problems, but it might be still worth integrating these changes on your side, they shouldn't hurt (famous last words...) and just might help.

vadz committed:
Use C++11 range for loops in wxModalDialogHook

No real changes, just make the code a bit less heavy.

vadz committed:
Add wxModalDialogHook::GetOpenCount()

It doesn't cost almost anything to provide this function, as we already
track showing and dismissing all modal dialogs in this class, and it can
(will) be useful, so add it.

vadz committed:
Remove wxGTK-specific wxOpenModalDialogsCount

Use wxModalDialogHook::GetOpenCount() added in the previous commit
instead.

This is more complete, as wxOpenModalDialogsCount was only updated when
showing generic and message box dialogs, while wxModalDialogHook is used
for all modal dialogs, including file open/save ones etc.

vadz committed:
Veto closing top level windows if any modal dialogs are shown

We can't always allow the window to close before the modal dialog is and
we don't know if it's safe to close it or not, in general, so err on the
side of the caution.

vadz committed:
Always send wxEVT_END_SESSION in wxMSW exactly once

We could skip generating this event completely if there was no main
application window any longer by the time we receive WM_ENDSESSION,
which was perfectly possible because WM_QUERYENDSESSION handler closes
all windows by default, making them "pending for destruction" and such
windows are not considered by GetMainTopWindow(), so it always returned
null pointer in this case.

Instead of comparing the window, use a separate flag indicating whether
we already had received this event, which will also be useful for other
purposes, see the upcoming commit.

vadz committed:
Avoid spurious error about unregistering classes on session end

We would always fail to unregister our window class because the window
using it was still open, so don't even try doing it -- it's not really
useful anyhow, because everything will be torn down soon by the system
in any case.

vadz commented:
I think the simplest way to make sure we never process this event twice is to not rely on wx event handling but just hard code what `wxApp::OnEndSession()` does in a function that will be called no matter what, i.e. whether the event is handled or skipped.

What do you think?

But the linked Poedit crash seems to happen due to a completely unexpected reentrancy into the message loop from inside `::DestroyWindow()` and I have no idea what to do about this, nothing here is going to help with this.

vadz committed:
Get rid of CppUnit boilerplate in modal dialog test cases

No real changes, just use CATCH macros directly.

vadz committed:
Don't destroy any windows during the session shutdown

Instead of skipping the destruction of just one top level window, as was
previously done in wxApp::OnEndSession(), don't destroy any of them once
we get WM_ENDSESSION in order to avoid dispatching any messages,
including nested WM_ENDSESSION, as it could apparently happen when
destroying at least some kinds of windows.

vadz committed:
Document that wxEVT_END_SESSION handler shouldn't use any UI

Also advise skipping the event.

vadz commented:
> Sorry, I'm not sure I follow?

I think this might be the case when the code could be more clear than my explanations, especially because I've simplified it to do only 2(i) from above because, after thinking more about this, I think this makes the rest unnecessary, i.e. it looks like this on its own should take care of the problems identified so far.

But please have a look at the latest version and tell me what do you think about it. TIA!

vslavik commented on src/msw/window.cpp:
```diff
@@ -4351,9 +4359,14 @@ bool wxWindowMSW::HandleEndSession(bool endSession, long logOff)
     if ( !endSession )
         return false;
```
This branch should be amended with `gs_gotEndSession = false;` so that any subsequent new attempts to end the session work correctly.

vadz replied:
We could do it, but I think it's unnecessary and potentially confusing because it contradicts the claim that once `gs_gotEndSession` is set to true the application is sure to exit. Am I missing some scenario in which you think we may get `WM_ENDSESSION(TRUE)` first followed by `WM_ENDSESSION(FALSE)` later?

vslavik replied:
Sorry, I got confused about `WM_ENDSESSION` vs `WM_QUERYENDSESSION`. You're right, sorry :(

vslavik commented:
> But please have a look at the latest version and tell me what do you think about it. TIA!

Looks reasonable to me (and better than https://github.com/wxWidgets/wxWidgets/pull/24903), thanks!

vadz commented:
I think we should merge this already and see if it helps, but, just for the record, I still have a couple of ideas about other small changes:

1. We should send `wxEVT_QUERY_END_SESSION` to the window itself in `wxWindow::HandleQueryEndSession()` instead of sending it to wxApp (of course, it will still end up in wxApp if the window doesn't handle it) and have a default handler for it in `wxTLW` which would call `Close()` to ensure that we close all TLWs and not just the main one.
1. If we can confirm that modern Windows versions kill the process once `WM_ENDSESSION` returns, I think we ought to call `exit()` from `wxWindow::HandleEndSession()` itself after the user-defined handler returns because it looks strictly better to execute the global objects destructors than to not do it.

Please let me know if you think it's worth adding either/both of those to this PR, this would be simple to do and I could do it now (unlike more extensive changes, like adding support for the restart manager or providing block reason, which would be definitely worth doing but for which I don't really have time).

Otherwise I'll just merge the current version relatively soon (with or without `gs_gotEndSession` reset discussed in a separate conversation above).

vslavik commented:
> Please let me know if you think it's worth adding either/both of those to this PR, this would be simple to do

IMHO they are worth doing, but maybe not so simple and not worth delaying this PR?

> We should send wxEVT_QUERY_END_SESSION to the window itself in wxWindow::HandleQueryEndSession() instead of sending it to wxApp

Per-window handling would be convenient to have, but:
1. I am not sure how `WM_QUERYENDSESSION` is sent, if to all TLWs right away or sequentially; MSDN also talks about sending the message to application, so this may be intentionally vague
2. I do know that the first non-vetoed message results in immediately going to the `WM_ENDSESSION` phase, i.e. handling at wxApp level and iterating over TLWs seems safer
3. It's used in wxOSX too (where the native API is application-level)

> If we can confirm that modern Windows versions kill the process once WM_ENDSESSION returns

 I'm not sure how to reliably confirm Windows behavior w.r.t. killing the process

vadz commented:
> Per-window handling would be convenient to have, but:
> 
>     1. I am not sure how `WM_QUERYENDSESSION` is sent, if to all TLWs right away or sequentially; MSDN also talks about sending the message to application, so this may be intentionally vague

I'm pretty sure it's to all TLWs, sequentially, just as all the other broadcast messages.
 
>     2. I do know that the first non-vetoed message results in immediately going to the `WM_ENDSESSION` phase, i.e. handling at wxApp level and iterating over TLWs seems safer
> 
>     3. It's used in wxOSX too (where the native API is application-level)

Oh, this is a good point, thanks.

So finally we should indeed just iterate over all TLWs ourselves in `wxApp::OnQueryEndSession()`. But we still need to find some way of avoiding generating it more than once and this looks tricky (we could use the same `gs_gotEndSession` when it is not vetoed, but not if it is vetoed...).

> > If we can confirm that modern Windows versions kill the process once WM_ENDSESSION returns
> 
> I'm not sure how to reliably confirm Windows behavior w.r.t. killing the process

In principle this is simple: just open a file, and write to it something from `wxEVT_END_SESSION` handler and from some global dtor (always flushing it after writing), then check what ends up in the file in different scenarios. Missing messages would be a reliable indicator of Windows having killed the process before this point is reached, IMO.

vslavik commented:
> But we still need to find some way of avoiding generating it more than once and this looks tricky (we could use the same gs_gotEndSession when it is not vetoed, but not if it is vetoed...).

I just use a 5s timeout in user code now. But this is actually where I got my wires crossed earlier: we could have a `g_gotQueryEndSession` that is reset back to false when handling `WM_ENDSESSION(FALSE)` (which is sent when the OS cancels its session-ending attempt).

vadz committed:
Make sure we generate wxEVT_QUERY_END_SESSION only once too

Because WM_QUERYENDSESSION is sent to all application TLWs, we could
generate this event multiple times, which was probably rather
unexpected, so don't do this.

Note that we rely on Windows sending us WM_ENDSESSION(FALSE) to reset
the flag set in WM_QUERYENDSESSION, but this does seem to happen in all
the current Windows versions.

vadz committed:
Close all windows in wxApp::OnQueryEndSession()

Other windows than the main one might have unsaved changes and want to
veto ending the session too.

vadz commented:
> I just use a 5s timeout in user code now. But this is actually where I got my wires crossed earlier: we could have a `g_gotQueryEndSession` that is reset back to false when handling `WM_ENDSESSION(FALSE)` (which is sent when the OS cancels its session-ending attempt).

Ah, I didn't realize we could use this, thanks.

I've implemented what (I believe) you proposed now, but didn't have time to test it because I have to leave now, so that's all from me for today. Thanks again for the review and advice!

vslavik commented:
> I've implemented what (I believe) you proposed now

Yes, that's what I meant, thanks! It seems to behave correctly in my testing. I was able to clarify this bit of behavior (at least on Win11, with rmlogotest):

> > I am not sure how `WM_QUERYENDSESSION` is sent, if to all TLWs right away or sequentially; MSDN also talks about sending the message to application, so this may be intentionally vague
>
> I'm pretty sure it's to all TLWs, sequentially, just as all the other broadcast messages.

When I have a modal window shown in response to the first `WM_QUERYENDSESSION`, asking user for confirmation, then if the user doesn't respond within 5 seconds, another `WM_QUERYENDSESSION` is sent to another TLW (and is handled by the inner event loop of the modal message box) even though the first one wasn't responded to yet. So it's not strictly sequential, at least with rmlogotest.

However, I wasn't able to trigger any bad behavior even though I tried: replying to the 2nd message with Allow, as this PR's code does, still resulted in WM_ENDSESSION(FALSE), presumably because the 1st message's handling took so long. This is strange, because it is seemingly contrary to the docs, but it's unclear to me if this is due to using rmlogotest or is "real" behavior.

vadz commented:
Thanks for testing!

I am going to merge this soon if there are no objections from anybody else.
